File: cpp/problem441.cpp
#include <mystd.h>
using namespace MYSTD;
using namespace std;

// By using mathematical formula for staircase sum(1...k) = k(k+1)/2,
// solution is simply a root-findting problem.
class Solution {
public:
    int arrangeCoins(int n) {
        long b = -1, e = 65536, c, dy;
        if( n >= e*(e-1)/2 ) return e-1;
        do{
            c = (b + e)/2; dy = c*(c+1)/2 - n;
            if( dy > 0 ) e = c;
            else if( dy < 0 ) b = c;
            else return c;
        }while( e > b+1 );
        return b;
    }
};

int main(int argc, char const *argv[]){
    Solution sol;
    // cases in Leetcode Problem
    cout << sol.arrangeCoins(5) << endl;
    cout << sol.arrangeCoins(8) << endl;
    // other extreme cases
    cout << sol.arrangeCoins(0) << endl;
    cout << sol.arrangeCoins(1) << endl;
    return 0;
}

File: java/problem441.java
class Solution {
    public int arrangeCoins(int n) {
        if(n==0)return 0;
        int end=n;int beg=0;int med;
        while(end > beg){
            med=(end+beg)/2;
            if((((double)med+1)*(double)med)>(2*(double)n)){
                end=med;
            }else if((((double)med+1)*(double)med)<(2*(double)n)){
                beg=med;
            }else{
                return med;
            }
            //System.out.println("beg:"+beg+"; end:"+end+"; med:"+med);
            if(end-beg<=100)break;
        }
        int i;
        for(i=beg;i<=end;i++){
            if(((double)i*((double)i+1))>=(2*(double)n))break;
        }
        return ((double)i*((double)i+1))==(2*(double)n)?i:i-1;
    }
}
public class problem441{
     public static void main(String args[]){
         Solution test=new Solution();
         System.out.println(test.arrangeCoins(1804289383));
        //int med=112768085;int n=1804289383;
        //System.out.println((((double)med+1)*(double)med)+";"+(2*(double)n));
    }
}